File: sdk/node/test_workspace_retrieval_real_deepseek.mjs
import assert from 'node:assert/strict'
import crypto from 'node:crypto'
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'
import mod from './index.js'

const fixture = JSON.parse(fs.readFileSync(
  new URL('../evaluation/workspace-retrieval-deepseek-v1.json', import.meta.url),
  'utf8',
))
assert.equal(fixture.schema_version, 1)
assert.equal(fixture.report_schema_version, 1)
const validateOnly = process.argv.includes('--validate-fixture')
const readyTimeoutMs = 10_000
const turnTimeoutMs = 240_000

function generatedCorpusFiles() {
  const files = fixture.corpus.source_files.map((file) => ({
    path: file.path,
    bytes: Buffer.from(file.content, 'utf8'),
    text: true,
  }))
  for (let index = 0; index < fixture.corpus.unrelated_file_count; index += 1) {
    const padded = String(index).padStart(2, '0')
    files.push({
      path: `src/unrelated_${padded}.rs`,
      bytes: Buffer.from(
        `pub fn unrelated_worker_${padded}(value: usize) -> usize { value + ${index} }\n`,
        'utf8',
      ),
      text: true,
    })
  }
  let boundary = ''
  for (let index = 0; index < fixture.corpus.boundary_filler_lines; index += 1) {
    boundary += `// deterministic chunk-boundary filler ${String(index).padStart(2, '0')}\n`
  }
  boundary += 'pub const MAX_PENDING_EMBED_BATCHES: usize = 8;\n\n'
  boundary += 'pub fn admits_batch(pending: usize) -> bool {\n'
  boundary += '    pending < MAX_PENDING_EMBED_BATCHES\n}\n'
  files.push({
    path: 'src/embedding_admission.rs',
    bytes: Buffer.from(boundary, 'utf8'),
    text: true,
  })
  files.push(...fixture.corpus.non_text_files.map((file) => ({
    path: file.path,
    bytes: Buffer.from(file.content, 'utf8'),
    text: false,
  })))
  return files.sort((left, right) => left.path.localeCompare(right.path))
}

function materializeCorpus(root) {
  const files = generatedCorpusFiles()
  const hash = crypto.createHash('sha256')
  for (const file of files) {
    const destination = path.join(root, ...file.path.split('/'))
    fs.mkdirSync(path.dirname(destination), { recursive: true })
    fs.writeFileSync(destination, file.bytes)
    hash.update(Buffer.from(file.path, 'utf8'))
    hash.update(Buffer.from([0]))
    hash.update(file.bytes)
    hash.update(Buffer.from([0]))
  }
  assert.equal(files.filter((file) => file.text).length, fixture.corpus.text_file_count)
  assert.equal(files.filter((file) => !file.text).length, fixture.corpus.non_text_file_count)
  return hash.digest('hex')
}

function stableBucket(text, buckets) {
  let hash = 2166136261
  for (const byte of Buffer.from(text, 'utf8')) {
    hash ^= byte
    hash = Math.imul(hash, 16777619) >>> 0
  }
  return hash % buckets
}

function vectorFor(id, text) {
  let axis
  if (id === fixture.embedding.query_id) {
    axis = fixture.tasks.findIndex((task) => text.trim() === task.query)
    assert.notEqual(axis, -1, `unexpected evaluation query: ${text}`)
  } else {
    axis = fixture.tasks.findIndex((task) => text.includes(task.expected_identifier))
    if (axis === -1) {
      axis = fixture.tasks.length + stableBucket(
        text,
        fixture.embedding.dimension - fixture.tasks.length,
      )
    }
  }
  const vector = Array(fixture.embedding.dimension).fill(0)
  vector[axis] = 1
  return vector
}

function evaluationProvider(counters) {
  return new mod.CallbackEmbeddingProvider(
    {
      provider: fixture.embedding.provider,
      model: fixture.embedding.model,
      revision: fixture.embedding.revision,
      dimension: fixture.embedding.dimension,
      normalization: 'unit',
    },
    async (request) => {
      assert.equal(request.signal instanceof AbortSignal, true)
      const query = request.inputs.every((input) => input.id === fixture.embedding.query_id)
      const documents = request.inputs.every((input) => input.id !== fixture.embedding.query_id)
      assert.equal(query || documents, true, 'document and query inputs must not share a batch')
      counters.requests += 1
      counters.queryRequests += query ? 1 : 0
      counters.documentRequests += documents ? 1 : 0
      for (const input of request.inputs) {
        counters.inputBytes += Buffer.byteLength(input.text, 'utf8')
        counters.queryInputs += input.id === fixture.embedding.query_id ? 1 : 0
        counters.documentInputs += input.id === fixture.embedding.query_id ? 0 : 1
        counters.nonTextInputs += input.text.includes('NON_TEXT_ASSET_SENTINEL') ? 1 : 0
      }
      await Promise.resolve()
      return {
        vectors: request.inputs.map((input) => ({
          id: input.id,
          values: vectorFor(input.id, input.text),
        })),
      }
    },
  )
}

function percentile(values, percentileValue) {
  if (values.length === 0) return 0
  const sorted = [...values].sort((left, right) => left - right)
  const rank = Math.max(0, Math.ceil(percentileValue * sorted.length) - 1)
  return sorted[rank]
}

function normalizedAnswer(text) {
  return text.trim().replace(/^`+|`+$/g, '').trim()
}

function runEventsToolCalls(events) {
  return events
    .filter((event) => event.type === 'tool_end')
    .map((event) => event.payload)
}

async function withTimeout(promise, label, timeoutMs) {
  let timer
  const timeout = new Promise((_, reject) => {
    timer = setTimeout(() => reject(new Error(`${label} timed out after ${timeoutMs}ms`)), timeoutMs)
  })
  return Promise.race([promise, timeout]).finally(() => clearTimeout(timer))
}

async function waitUntilReady(session) {
  const started = Date.now()
  let status = session.workspaceRetrievalStatus()
  while (status.phase === 'building' && Date.now() - started < readyTimeoutMs) {
    await new Promise((resolve) => setTimeout(resolve, 10))
    status = session.workspaceRetrievalStatus()
  }
  assert.equal(status.phase, 'ready', JSON.stringify(status))
  return { status, elapsedMs: Date.now() - started }
}

function assertReadyStatus(status, counters) {
  assert.equal(status.coverageBps, 10_000)
  assert.equal(status.eligibleFiles, fixture.corpus.text_file_count)
  assert.equal(status.indexedFiles, fixture.corpus.text_file_count)
  assert.equal(status.indexedChunks, fixture.corpus.expected_chunk_count)
  assert.equal(status.failedFiles, 0)
  assert.equal(status.vectorRecords, fixture.corpus.expected_chunk_count)
  assert.equal(status.batching.documentInputs, fixture.corpus.expected_chunk_count)
  assert.equal(status.batching.documentProviderRequests, 1)
  assert.equal(status.batching.batchLimitLowerBound, 1)
  assert.equal(status.batching.nonTextInputs, 0)
  assert.notEqual(status.batching.timeToFirstReadyMs, undefined)
  assert.equal(counters.documentRequests, 1)
  assert.equal(counters.documentInputs, fixture.corpus.expected_chunk_count)
  assert.equal(counters.nonTextInputs, 0)
}

function taskPrompt(task) {
  return `Inspect the search tool schema. Make exactly one search call and no other tool call. Use query exactly: ${task.query}. Set path to '.', include to '*.rs', limit to 5, and mode to 'hybrid'. After the result, return exactly the Rust function or constant declaration name that directly answers the query and is supported by the evidence, or NOT_FOUND when no relevant declaration is present. Never return a path, file stem, module name, prose, or Markdown.`
}

function normalizedBatching(batching) {
  return {
    documentInputs: batching.documentInputs,
    documentTextBytes: batching.documentTextBytes,
    documentBatches: batching.documentBatches,
    documentProviderRequests: batching.documentProviderRequests,
    batchLimitLowerBound: batching.batchLimitLowerBound,
    inputLimitFlushes: batching.inputLimitFlushes,
    textByteLimitFlushes: batching.textByteLimitFlushes,
    vectorByteLimitFlushes: batching.vectorByteLimitFlushes,
    generationCompleteFlushes: batching.generationCompleteFlushes,
    timeToFirstReadyMs: batching.timeToFirstReadyMs,
    nonTextInputs: batching.nonTextInputs,
  }
}

async function runTask(agent, task, ordinal) {
  const workspace = fs.mkdtempSync(path.join(os.tmpdir(), 'a3s-node-wsr-real-'))
  const corpusDigest = materializeCorpus(workspace)
  assert.equal(corpusDigest, fixture.corpus.expected_digest)
  const counters = {
    requests: 0,
    documentRequests: 0,
    queryRequests: 0,
    documentInputs: 0,
    queryInputs: 0,
    inputBytes: 0,
    nonTextInputs: 0,
  }
  const provider = evaluationProvider(counters)
  const chunking = new mod.RecursiveWorkspaceChunkingStrategy(
    fixture.chunking.target_bytes,
    fixture.chunking.overlap_bytes,
    fixture.chunking.separators,
  )
  const reranker = new mod.DeterministicWorkspaceReranker()
  const retrieval = new mod.WorkspaceRetrievalOptions(provider, reranker, chunking)
  const constructionStarted = Date.now()
  const session = await agent.sessionAsync(workspace, {
    sessionId: `wsr-sdk-node-${ordinal}`,
    model: fixture.chat_model,
    planningMode: 'disabled',
    goalTracking: false,
    permissionPolicy: { allow: ['search(*)'], defaultDecision: 'deny' },
    guidelines: 'This is a deterministic repository retrieval evaluation. Follow the requested one-tool protocol exactly. Never guess an identifier that is absent from the tool evidence.',
    maxParseRetries: 1,
    maxToolRounds: 2,
    manualDelegationEnabled: false,
    autoDelegation: { enabled: false },
    temperature: 0,
    workspaceRetrieval: retrieval,
  })
  const sessionConstructionMs = Date.now() - constructionStarted
  try {
    const ready = await waitUntilReady(session)
    assertReadyStatus(ready.status, counters)
    const turnStarted = Date.now()
    const result = await withTimeout(session.send(taskPrompt(task)), task.name, turnTimeoutMs)
    const turnElapsedMs = Date.now() - turnStarted
    const runs = await session.runs()
    assert.equal(runs.length, 1)
    assert.equal(runs[0].status, 'completed')
    const events = await session.runEvents(runs[0].id)
    const calls = runEventsToolCalls(events)
    const call = calls[0]
    const args = call?.args ?? {}
    const metadata = call?.metadata ?? {}
    const results = Array.isArray(metadata.results) ? metadata.results : []
    const expectedIndex = results.findIndex((entry) => entry.path === task.expected_path)
    const expectedPathRank = expectedIndex === -1 ? null : expectedIndex + 1
    const protocolOk = result.toolCallsCount === 1
      && calls.length === 1
      && call.name === 'search'
      && call.exit_code === 0
      && args.query === task.query
      && args.path === '.'
      && args.include === '*.rs'
      && args.limit === 5
      && args.mode === 'hybrid'
    const completionCorrect = normalizedAnswer(result.text) === task.expected_identifier
    assert.equal(protocolOk, true, JSON.stringify(calls))
    assert.equal(completionCorrect, true, result.text)
    assert.notEqual(expectedPathRank, null, JSON.stringify(results))
    assert.equal(metadata.rerank?.requestedMode, fixture.rerank.requested_mode)
    assert.equal(metadata.rerank?.appliedMode, fixture.rerank.requested_mode)
    assert.equal(metadata.algorithm, fixture.rerank.algorithm)
    assert.equal(counters.queryRequests, 1)
    assert.equal(counters.queryInputs, 1)
    const closeStarted = Date.now()
    await session.closeAsync()
    const closeMs = Date.now() - closeStarted
    const closed = session.workspaceRetrievalStatus()
    assert.equal(closed.phase, 'closed')
    assert.equal(closed.vectorRecords, 0)
    assert.equal(closed.vectorBytes, 0)
    return {
      task: task.name,
      completionCorrect,
      toolProtocolOk: protocolOk,
      expectedPathRank,
      resultCount: results.length,
      algorithm: metadata.algorithm,
      rerankRequestedMode: metadata.rerank?.requestedMode,
      rerankAppliedMode: metadata.rerank?.appliedMode,
      sessionConstructionMs,
      indexReadyMs: ready.elapsedMs,
      turnElapsedMs,
      closeMs,
      promptTokens: result.promptTokens,
      completionTokens: result.completionTokens,
      totalTokens: result.totalTokens,
      phase: ready.status.phase,
      coverageBps: ready.status.coverageBps,
      eligibleFiles: ready.status.eligibleFiles,
      indexedFiles: ready.status.indexedFiles,
      indexedChunks: ready.status.indexedChunks,
      vectorRecords: ready.status.vectorRecords,
      vectorBytes: ready.status.vectorBytes,
      batching: normalizedBatching(ready.status.batching),
      provider: counters,
      releasedAfterClose: true,
    }
  } finally {
    if (session.workspaceRetrievalStatus().phase !== 'closed') await session.closeAsync()
    fs.rmSync(workspace, { recursive: true, force: true })
  }
}

function summarize(runs) {
  const ranks = runs.map((run) => run.expectedPathRank)
  const relevant = ranks.filter((rank) => rank !== null && rank <= 5).length
  const returned = runs.reduce((sum, run) => sum + run.resultCount, 0)
  const lowerBound = runs.reduce((sum, run) => sum + run.batching.batchLimitLowerBound, 0)
  const documentRequests = runs.reduce(
    (sum, run) => sum + run.batching.documentProviderRequests,
    0,
  )
  return {
    taskAccuracy: runs.filter((run) => run.completionCorrect).length / runs.length,
    toolProtocolRate: runs.filter((run) => run.toolProtocolOk).length / runs.length,
    precisionAt5: relevant / (runs.length * 5),
    returnedResultPrecision: relevant / returned,
    recallAt5: relevant / runs.length,
    mrr: ranks.reduce((sum, rank) => sum + (rank === null ? 0 : 1 / rank), 0) / runs.length,
    ndcgAt5: ranks.reduce(
      (sum, rank) => sum + (rank === null || rank > 5 ? 0 : 1 / Math.log2(rank + 1)),
      0,
    ) / runs.length,
    documentRequestAmplification: documentRequests / lowerBound,
    meanReturnedResults: returned / runs.length,
    sessionConstructionP50Ms: percentile(runs.map((run) => run.sessionConstructionMs), 0.5),
    sessionConstructionP95Ms: percentile(runs.map((run) => run.sessionConstructionMs), 0.95),
    indexReadyP50Ms: percentile(runs.map((run) => run.indexReadyMs), 0.5),
    indexReadyP95Ms: percentile(runs.map((run) => run.indexReadyMs), 0.95),
    timeToFirstReadyP50Ms: percentile(
      runs.map((run) => run.batching.timeToFirstReadyMs),
      0.5,
    ),
    timeToFirstReadyP95Ms: percentile(
      runs.map((run) => run.batching.timeToFirstReadyMs),
      0.95,
    ),
    turnP50Ms: percentile(runs.map((run) => run.turnElapsedMs), 0.5),
    turnP95Ms: percentile(runs.map((run) => run.turnElapsedMs), 0.95),
    closeP50Ms: percentile(runs.map((run) => run.closeMs), 0.5),
    closeP95Ms: percentile(runs.map((run) => run.closeMs), 0.95),
    totalTokens: runs.reduce((sum, run) => sum + run.totalTokens, 0),
    nonTextProviderInputs: runs.reduce((sum, run) => sum + run.provider.nonTextInputs, 0),
    releasedAfterCloseRate: runs.filter((run) => run.releasedAfterClose).length / runs.length,
  }
}

const validationRoot = fs.mkdtempSync(path.join(os.tmpdir(), 'a3s-node-wsr-fixture-'))
try {
  const digest = materializeCorpus(validationRoot)
  if (fixture.corpus.expected_digest !== 'PENDING') {
    assert.equal(digest, fixture.corpus.expected_digest)
  }
  if (validateOnly) {
    console.log(`Node workspace retrieval fixture validated: ${digest}`)
    process.exitCode = 0
  } else {
    const evaluationRoot = process.env.A3S_REAL_EVAL_ROOT
    assert.ok(evaluationRoot, 'A3S_REAL_EVAL_ROOT must point to the a3s monorepo root')
    const configPath = path.resolve(evaluationRoot, '.a3s', 'config.acl')
    assert.equal(fs.existsSync(configPath), true, 'repository .a3s/config.acl is required')
    const agent = await mod.Agent.create(configPath)
    try {
      const runs = []
      for (const [ordinal, task] of fixture.tasks.entries()) {
        runs.push(await runTask(agent, task, ordinal))
      }
      const summary = summarize(runs)
      assert.equal(summary.taskAccuracy, 1)
      assert.equal(summary.toolProtocolRate, 1)
      assert.equal(summary.recallAt5, 1)
      assert.ok(summary.documentRequestAmplification <= 1.1)
      assert.equal(summary.nonTextProviderInputs, 0)
      assert.equal(summary.releasedAfterCloseRate, 1)
      const report = {
        schemaVersion: fixture.report_schema_version,
        fixtureId: fixture.fixture_id,
        fixtureDigest: fixture.corpus.expected_digest,
        sdk: 'node',
        chatModel: fixture.chat_model,
        chunking: fixture.chunking,
        rerank: fixture.rerank,
        summary,
        runs,
        allGatesPassed: true,
      }
      console.log(`WSR_SDK_DEEPSEEK_EVAL=${JSON.stringify(report)}`)
    } finally {
      await agent.close()
    }
  }
} finally {
  fs.rmSync(validationRoot, { recursive: true, force: true })
}

File: sdk/python/tests/test_workspace_retrieval_real_deepseek.py
"""Cross-SDK real-DeepSeek qualification for Python workspace retrieval."""

from __future__ import annotations

import asyncio
import json
import os
import shutil
import sys
import tempfile
import time
import unittest
from pathlib import Path
from typing import Any, cast

from workspace_retrieval_eval_fixture import (
    FIXTURE,
    materialize_corpus,
    validate_fixture_contract,
)
from workspace_retrieval_deepseek_report import summarize

from a3s_code import (
    Agent,
    CallbackEmbeddingProvider,
    DeterministicWorkspaceReranker,
    EmbeddingBatchRequest,
    EmbeddingBatchResponse,
    PermissionPolicy,
    RecursiveWorkspaceChunkingStrategy,
    SessionOptions,
    WorkspaceRetrievalOptions,
    WorkspaceRetrievalStatus,
)


READY_TIMEOUT_SECONDS = 10.0
TURN_TIMEOUT_SECONDS = 240.0


def stable_bucket(text: str, buckets: int) -> int:
    value = 2166136261
    for byte in text.encode("utf-8"):
        value ^= byte
        value = (value * 16777619) & 0xFFFFFFFF
    return value % buckets


def vector_for(input_id: str, text: str) -> list[float]:
    tasks = FIXTURE["tasks"]
    if input_id == FIXTURE["embedding"]["query_id"]:
        matches = [index for index, task in enumerate(tasks) if text.strip() == task["query"]]
        assert len(matches) == 1, f"unexpected evaluation query: {text}"
        axis = matches[0]
    else:
        matches = [
            index
            for index, task in enumerate(tasks)
            if task["expected_identifier"] in text
        ]
        axis = (
            matches[0]
            if matches
            else len(tasks)
            + stable_bucket(text, FIXTURE["embedding"]["dimension"] - len(tasks))
        )
    vector = [0.0] * FIXTURE["embedding"]["dimension"]
    vector[axis] = 1.0
    return vector


def evaluation_provider(counters: dict[str, int]) -> CallbackEmbeddingProvider:
    async def embed(request: EmbeddingBatchRequest) -> EmbeddingBatchResponse:
        inputs = request["inputs"]
        query = all(item["id"] == FIXTURE["embedding"]["query_id"] for item in inputs)
        documents = all(item["id"] != FIXTURE["embedding"]["query_id"] for item in inputs)
        assert query or documents, "document and query inputs must not share a batch"
        counters["requests"] += 1
        counters["query_requests"] += int(query)
        counters["document_requests"] += int(documents)
        for item in inputs:
            is_query = item["id"] == FIXTURE["embedding"]["query_id"]
            counters["input_bytes"] += len(item["text"].encode("utf-8"))
            counters["query_inputs"] += int(is_query)
            counters["document_inputs"] += int(not is_query)
            counters["non_text_inputs"] += int(
                "NON_TEXT_ASSET_SENTINEL" in item["text"]
            )
        await asyncio.sleep(0)
        return {
            "vectors": [
                {
                    "id": item["id"],
                    "values": vector_for(item["id"], item["text"]),
                }
                for item in inputs
            ]
        }

    embedding = FIXTURE["embedding"]
    return CallbackEmbeddingProvider(
        embedding["provider"],
        embedding["model"],
        embedding["dimension"],
        embed,
        revision=embedding["revision"],
        normalization="unit",
    )


async def wait_until_ready(session: Any) -> tuple[WorkspaceRetrievalStatus, int]:
    started = time.monotonic()
    status = cast(WorkspaceRetrievalStatus, session.workspace_retrieval_status())
    while status["phase"] == "building":
        if time.monotonic() - started >= READY_TIMEOUT_SECONDS:
            raise TimeoutError("workspace retrieval did not become ready")
        await asyncio.sleep(0.01)
        status = cast(WorkspaceRetrievalStatus, session.workspace_retrieval_status())
    assert status["phase"] == "ready", status
    return status, int((time.monotonic() - started) * 1000)


def assert_ready_status(
    status: WorkspaceRetrievalStatus, counters: dict[str, int]
) -> None:
    corpus = FIXTURE["corpus"]
    batching = status["batching"]
    assert status["coverage_bps"] == 10_000
    assert status["eligible_files"] == corpus["text_file_count"]
    assert status["indexed_files"] == corpus["text_file_count"]
    assert status["indexed_chunks"] == corpus["expected_chunk_count"]
    assert status["failed_files"] == 0
    assert status["vector_records"] == corpus["expected_chunk_count"]
    assert batching["document_inputs"] == corpus["expected_chunk_count"]
    assert batching["document_provider_requests"] == 1
    assert batching["batch_limit_lower_bound"] == 1
    assert batching["non_text_inputs"] == 0
    assert batching["time_to_first_ready_ms"] is not None
    assert counters["document_requests"] == 1
    assert counters["document_inputs"] == corpus["expected_chunk_count"]
    assert counters["non_text_inputs"] == 0


def task_prompt(task: dict[str, str]) -> str:
    return (
        "Inspect the search tool schema. Make exactly one search call and no other "
        f"tool call. Use query exactly: {task['query']}. Set path to '.', include "
        "to '*.rs', limit to 5, and mode to 'hybrid'. After the result, return "
        "exactly the Rust function or constant declaration name that directly "
        "answers the query and is supported by the evidence, or NOT_FOUND when "
        "no relevant declaration is present. Never return a path, file stem, "
        "module name, prose, or Markdown."
    )


def normalized_answer(text: str) -> str:
    return text.strip().strip("`").strip()


async def run_task(agent: Any, task: dict[str, str], ordinal: int) -> dict[str, Any]:
    workspace = Path(tempfile.mkdtemp(prefix="a3s-python-wsr-real-"))
    session = None
    try:
        corpus_digest = materialize_corpus(workspace)
        assert corpus_digest == FIXTURE["corpus"]["expected_digest"]
        counters = {
            "requests": 0,
            "document_requests": 0,
            "query_requests": 0,
            "document_inputs": 0,
            "query_inputs": 0,
            "input_bytes": 0,
            "non_text_inputs": 0,
        }
        chunking = RecursiveWorkspaceChunkingStrategy(
            FIXTURE["chunking"]["target_bytes"],
            FIXTURE["chunking"]["overlap_bytes"],
            FIXTURE["chunking"]["separators"],
        )
        retrieval = WorkspaceRetrievalOptions(
            evaluation_provider(counters),
            DeterministicWorkspaceReranker(),
            chunking,
        )
        options = SessionOptions()
        options.session_id = f"wsr-sdk-python-{ordinal}"
        options.model = FIXTURE["chat_model"]
        options.planning_mode = "disabled"
        options.goal_tracking = False
        options.permission_policy = PermissionPolicy(
            allow=["search(*)"], default_decision="deny"
        )
        options.guidelines = (
            "This is a deterministic repository retrieval evaluation. Follow the "
            "requested one-tool protocol exactly. Never guess an identifier that "
            "is absent from the tool evidence."
        )
        options.max_parse_retries = 1
        options.max_tool_rounds = 2
        options.manual_delegation_enabled = False
        options.auto_parallel = False
        options.temperature = 0.0
        options.workspace_retrieval = retrieval
        construction_started = time.monotonic()
        session = await agent.session_async(str(workspace), options)
        session_construction_ms = int((time.monotonic() - construction_started) * 1000)
        status, index_ready_ms = await wait_until_ready(session)
        assert_ready_status(status, counters)
        turn_started = time.monotonic()
        result = await asyncio.wait_for(
            session.send_async(task_prompt(task)), timeout=TURN_TIMEOUT_SECONDS
        )
        turn_elapsed_ms = int((time.monotonic() - turn_started) * 1000)
        runs = await session.runs_async()
        assert len(runs) == 1
        assert runs[0]["status"] == "completed"
        events = await session.run_events_async(runs[0]["id"])
        calls = [event["payload"] for event in events if event["type"] == "tool_end"]
        call = calls[0] if calls else {}
        args = call.get("args") or {}
        metadata = call.get("metadata") or {}
        results = metadata.get("results") or []
        expected_indexes = [
            index
            for index, entry in enumerate(results)
            if entry.get("path") == task["expected_path"]
        ]
        expected_path_rank = expected_indexes[0] + 1 if expected_indexes else None
        protocol_ok = (
            result.tool_calls_count == 1
            and len(calls) == 1
            and call.get("name") == "search"
            and call.get("exit_code") == 0
            and args.get("query") == task["query"]
            and args.get("path") == "."
            and args.get("include") == "*.rs"
            and args.get("limit") == 5
            and args.get("mode") == "hybrid"
        )
        completion_correct = (
            normalized_answer(result.text) == task["expected_identifier"]
        )
        assert protocol_ok, calls
        assert completion_correct, result.text
        assert expected_path_rank is not None, results
        rerank = metadata.get("rerank") or {}
        assert rerank.get("requestedMode") == FIXTURE["rerank"]["requested_mode"]
        assert rerank.get("appliedMode") == FIXTURE["rerank"]["requested_mode"]
        assert metadata.get("algorithm") == FIXTURE["rerank"]["algorithm"]
        assert counters["query_requests"] == 1
        assert counters["query_inputs"] == 1
        close_started = time.monotonic()
        await session.close_async()
        close_ms = int((time.monotonic() - close_started) * 1000)
        closed = cast(WorkspaceRetrievalStatus, session.workspace_retrieval_status())
        assert closed["phase"] == "closed"
        assert closed["vector_records"] == 0
        assert closed["vector_bytes"] == 0
        return {
            "task": task["name"],
            "completionCorrect": completion_correct,
            "toolProtocolOk": protocol_ok,
            "expectedPathRank": expected_path_rank,
            "resultCount": len(results),
            "algorithm": metadata.get("algorithm"),
            "rerankRequestedMode": rerank.get("requestedMode"),
            "rerankAppliedMode": rerank.get("appliedMode"),
            "sessionConstructionMs": session_construction_ms,
            "indexReadyMs": index_ready_ms,
            "turnElapsedMs": turn_elapsed_ms,
            "closeMs": close_ms,
            "promptTokens": result.prompt_tokens,
            "completionTokens": result.completion_tokens,
            "totalTokens": result.total_tokens,
            "phase": status["phase"],
            "coverageBps": status["coverage_bps"],
            "eligibleFiles": status["eligible_files"],
            "indexedFiles": status["indexed_files"],
            "indexedChunks": status["indexed_chunks"],
            "vectorRecords": status["vector_records"],
            "vectorBytes": status["vector_bytes"],
            "batching": {
                "documentInputs": status["batching"]["document_inputs"],
                "documentTextBytes": status["batching"]["document_text_bytes"],
                "documentBatches": status["batching"]["document_batches"],
                "documentProviderRequests": status["batching"][
                    "document_provider_requests"
                ],
                "batchLimitLowerBound": status["batching"]["batch_limit_lower_bound"],
                "inputLimitFlushes": status["batching"]["input_limit_flushes"],
                "textByteLimitFlushes": status["batching"][
                    "text_byte_limit_flushes"
                ],
                "vectorByteLimitFlushes": status["batching"][
                    "vector_byte_limit_flushes"
                ],
                "generationCompleteFlushes": status["batching"][
                    "generation_complete_flushes"
                ],
                "timeToFirstReadyMs": status["batching"]["time_to_first_ready_ms"],
                "nonTextInputs": status["batching"]["non_text_inputs"],
            },
            "provider": {
                "requests": counters["requests"],
                "documentRequests": counters["document_requests"],
                "queryRequests": counters["query_requests"],
                "documentInputs": counters["document_inputs"],
                "queryInputs": counters["query_inputs"],
                "inputBytes": counters["input_bytes"],
                "nonTextInputs": counters["non_text_inputs"],
            },
            "releasedAfterClose": True,
        }
    finally:
        if session is not None:
            current = cast(WorkspaceRetrievalStatus, session.workspace_retrieval_status())
            if current["phase"] != "closed":
                await session.close_async()
        shutil.rmtree(workspace, ignore_errors=True)


async def evaluate() -> dict[str, Any]:
    evaluation_root = os.environ.get("A3S_REAL_EVAL_ROOT")
    if not evaluation_root:
        raise RuntimeError("A3S_REAL_EVAL_ROOT must point to the a3s monorepo root")
    config_path = Path(evaluation_root, ".a3s", "config.acl").resolve()
    if not config_path.is_file():
        raise RuntimeError("repository .a3s/config.acl is required")
    agent = await Agent.create_async(str(config_path))
    try:
        runs = [
            await run_task(agent, task, ordinal)
            for ordinal, task in enumerate(FIXTURE["tasks"])
        ]
        summary = summarize(runs)
        assert summary["taskAccuracy"] == 1
        assert summary["toolProtocolRate"] == 1
        assert summary["recallAt5"] == 1
        assert summary["documentRequestAmplification"] <= 1.1
        assert summary["nonTextProviderInputs"] == 0
        assert summary["releasedAfterCloseRate"] == 1
        return {
            "schemaVersion": FIXTURE["report_schema_version"],
            "fixtureId": FIXTURE["fixture_id"],
            "fixtureDigest": FIXTURE["corpus"]["expected_digest"],
            "sdk": "python",
            "chatModel": FIXTURE["chat_model"],
            "chunking": FIXTURE["chunking"],
            "rerank": FIXTURE["rerank"],
            "summary": summary,
            "runs": runs,
            "allGatesPassed": True,
        }
    finally:
        await agent.close_async()


def test_workspace_retrieval_real_fixture_contract() -> None:
    validate_fixture_contract()


def test_workspace_retrieval_real_deepseek() -> None:
    if not os.environ.get("A3S_REAL_EVAL_ROOT"):
        raise unittest.SkipTest("set A3S_REAL_EVAL_ROOT to run the DeepSeek evaluation")
    report = asyncio.run(evaluate())
    print(f"WSR_SDK_DEEPSEEK_EVAL={json.dumps(report, ensure_ascii=False, separators=(',', ':'))}")


if __name__ == "__main__":
    if "--validate-fixture" in sys.argv:
        print(f"Python workspace retrieval fixture validated: {validate_fixture_contract()}")
    else:
        report = asyncio.run(evaluate())
        print(f"WSR_SDK_DEEPSEEK_EVAL={json.dumps(report, ensure_ascii=False, separators=(',', ':'))}")
